### PR TITLE
Adding/Editing Generators Documentation

### PR DESCRIPTION
**Pull request by jpreese**
For #264 

Rough draft! Please ding me on spelling/grammatical issues, structure, content, etc.

Once we finalize this, I'll go through and spin up new tickets for each generator we have remaining and reference this document on those tickets. I'd like them to be worked on during hacktoberfest, so please show some restraint before then 💟

### CODE REVIEW

File: docs/GENERATORS.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Test Generators
+
+Test generators allow tracks to generate tests automatically without having to write them ourselves. Each test generator reads from the exercises `canonical data`, which defines the name of the test, its inputs, and outputs. You can read more about exercisms approach to test suites [here](https://github.com/exercism/docs/blob/master/language-tracks/exercises/anatomy/test-suites.md).
```
Comment (ErikSchierboom) [inline docs/GENERATORS.md]: exercisms -> exercism's

File: docs/GENERATORS.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+- Canonical Data - Represents the entire test suite.
+- Canonical Data Case - A representation of a single test case.
+- Description - The name of the test.
+- Property - The method to be called when running the test.
+- Input - The input for the test case.
+- Expected - The expected value when running the test case.
+
+## Adding A Simple Generator
+Adding a generator file is quite simple. Simply add a new file to the generators folder with the name of the test (in CamelCase), and extend the Exercise abstract class.
```
Comment (ErikSchierboom) [inline docs/GENERATORS.md]: "name of the test" -> "name of the exercise"

Comment (ErikSchierboom) [inline docs/GENERATORS.md]: Please enclose Exercise with backticks to make it render as code.

File: docs/GENERATORS.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+```csharp
+namespace Generators.Exercises
+{
+    public class Bob : Exercise
+    {
+    }
+}
+```
+
+This is a fully working generator, no other code needs to be written. However, it's simplicity stems from the fact that the test suite and the program itself is reletively trivial.
```
Comment (ErikSchierboom) [inline docs/GENERATORS.md]: reletively -> relatively

File: docs/GENERATORS.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+## Adding A Complex Generator
+
+A more *complex* generator would be the ComplexNumbers generator found [here](https://github.com/exercism/csharp/blob/master/generators/Exercises/ComplexNumbers.cs).
+
+The `Exercise` abstract class currently exposes five methods that are used for overriding the default behavior when generating an exercise.
+
+### void UpdateCanonicalData(CanonicalData canonicalData)
+Update the canonical data for a given test. 
+
+The most common use for this override is to iterate through each of the canonical data's case.
```
Comment (ErikSchierboom) [inline docs/GENERATORS.md]: iterate through -> iterate over

Comment (ErikSchierboom) [reply docs/GENERATORS.md]: canonical data's case -> canonical data cases

File: docs/GENERATORS.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+Here the **Assert** is being overridden, as the Assert needs to call additional functions, but only if the property is consistency. Otherwise, render the assert as usual.
+
+## Updating Existing Files
+It is possible that an existing exercise does not match the canonical data. It is OK to update the exercise stub and/or the exercise example to follow the canonical data! An example might be that an exercise is named SumOfMultiples, but the SumOfMultiples.cs and Example.cs files both use `Multiples` as the name of the class.
+
+Also, if you find an issue with one of the existing generators or test suites simply open up the generator that you would like to update, make your changes, and then run the generators.
+
+## Running The Generators
+This repository is coded against .NET Core. To run all of the generators all you need to do is run the following command in the generators directory:
```
Comment (ErikSchierboom) [inline docs/GENERATORS.md]: Maybe add a link to the .NET Core installation instructions?

File: docs/GENERATORS.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+`dotnet run generators`
```
Comment (ErikSchierboom) [inline docs/GENERATORS.md]: You only need `dotnet run` IIRC.

File: docs/GENERATORS.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+This command will take all of the generators that are in the `Exercises` folder, and generate all of the test cases for that exercise. We use reflection to get all of the exercises, so if you are adding a new test, the test will be automatically included when running the generator.
+
+Once the generator has been ran, you can view the output of your generation by navigating to the test file for that exercise. As an example, the test suite for the Bob exercise can be found at:
```
Comment (ErikSchierboom) [inline docs/GENERATORS.md]: has been ran -> has been run

File: docs/GENERATORS.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+`csharp/exercises/bob/BobTest.cs`
```
Comment (ErikSchierboom) [inline docs/GENERATORS.md]: I think you can omit the `csharp/` part, as people are probably already in the project's directory.

**Comment by jpreese** [comment]
Alright I've made the requested changes as well as a couple more of my own.

**Comment by ErikSchierboom** [comment]
Thanks a lot!

File: docs/GENERATORS.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+protected override HashSet<string> AddAdditionalNamespaces()
+{
+  return new HashSet<string>()
+  {
+    typeof(Dictionary<char, int>).Namespace
+  };
+}
+```
+
+This snippet would add the namespace that the `Dictionary<char,int>` collection lives in (`System.Collections.Generic`).
```
Comment (robkeim) [inline docs/GENERATORS.md]: Minor: missing space between `char,` and `int`

File: docs/GENERATORS.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+An example of this is the [RunLengthEncoding](https://github.com/exercism/csharp/blob/master/generators/Exercises/RunLengthEncoding.cs) test.
+
+Here the **Assert** is being overridden. The assert needs to call additional functions, but only if the property is `consistency`. Otherwise, render the assert as usual.
+
+## Updating Existing Files
+It is possible that an existing exercise does not match the canonical data. It is OK to update the exercise stub and/or the exercise example to follow the canonical data! An example might be that an exercise is named SumOfMultiples, but the SumOfMultiples.cs and Example.cs files both use `Multiples` as the name of the class.
+
+Also, if you find an issue with one of the existing generators or test suites simply open up the generator that you would like to update, make your changes, and then run the generators.
+
+## Running The Generators
```
Comment (robkeim) [inline docs/GENERATORS.md]: Is there a way currently to run one one exercise? I don't think so, but if so, it's worth calling out.

Comment (ErikSchierboom) [reply docs/GENERATORS.md]: @robkeim There is, by using the `-e <exercise>` command-line parameter.

Comment (robkeim) [reply docs/GENERATORS.md]: Nice! I think that's worth calling out in the md then.

**Pull request merged**